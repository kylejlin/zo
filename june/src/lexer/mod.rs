use crate::token::*;

use string_parser::get_string_value;
mod string_parser;

#[cfg(test)]
mod tests;

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct LexError(ByteIndex, ByteIndex);

pub fn lex(s: &str) -> Result<Vec<Token>, LexError> {
    Lexer::new(s).lex()
}

struct Lexer<'a> {
    src: &'a str,
    out: Vec<Token>,
    state: State,
}

#[derive(Clone, Copy)]
enum State {
    Main,
    Word { start: ByteIndex, byte_len: usize },
    String { start: ByteIndex, byte_len: usize },
    Slash(ByteIndex),
    Dash(ByteIndex),
    SingleLineComment,
}

impl<'a> Lexer<'a> {
    fn new(src: &'a str) -> Self {
        Lexer {
            src,
            out: Vec::new(),
            state: State::Main,
        }
    }
}

impl Lexer<'_> {
    fn lex(mut self) -> Result<Vec<Token>, LexError> {
        for (current_index, curent) in self.src.char_indices() {
            self.handle_char(ByteIndex(current_index), curent)?;
        }
        self.finish_pending_state_and_reset()?;
        Ok(self.out)
    }

    fn handle_char(&mut self, current_index: ByteIndex, current: char) -> Result<(), LexError> {
        match self.state {
            State::Main => self.handle_char_assuming_state_is_main(current_index, current),
            State::Word { start, byte_len } => {
                self.handle_char_assuming_state_is_word(current_index, current, start, byte_len)
            }
            State::String { start, byte_len } => {
                self.handle_char_assuming_state_is_string(current, start, byte_len)
            }
            State::Slash(_) => self.handle_char_assuming_state_is_slash(current_index, current),
            State::Dash(_) => self.handle_char_assuming_state_is_dash(current_index, current),
            State::SingleLineComment => {
                self.handle_char_assuming_state_is_single_line_comment(current)
            }
        }
    }

    fn handle_char_assuming_state_is_main(
        &mut self,
        current_index: ByteIndex,
        current: char,
    ) -> Result<(), LexError> {
        match current {
            ' ' | '\t' | '\n' => {}
            '"' => {
                self.state = State::String {
                    start: current_index,
                    byte_len: '"'.len_utf8(),
                }
            }
            'a'..='z' | 'A'..='Z' | '_' | '0'..='9' => {
                self.state = State::Word {
                    start: current_index,
                    byte_len: 1,
                }
            }
            '(' => self.out.push(Token::LParen(current_index)),
            ')' => self.out.push(Token::RParen(current_index)),
            '[' => self.out.push(Token::LSquare(current_index)),
            ']' => self.out.push(Token::RSquare(current_index)),
            '=' => self.out.push(Token::Eq(current_index)),
            ':' => self.out.push(Token::Colon(current_index)),
            ',' => self.out.push(Token::Comma(current_index)),
            '^' => self.out.push(Token::Caret(current_index)),
            '-' => self.state = State::Dash(current_index),
            '/' => self.state = State::Slash(current_index),
            _ => {
                return Err(LexError(
                    current_index,
                    ByteIndex(current_index.0 + current.len_utf8()),
                ))
            }
        }
        Ok(())
    }

    fn handle_char_assuming_state_is_word(
        &mut self,
        current_index: ByteIndex,
        current: char,
        start: ByteIndex,
        byte_len: usize,
    ) -> Result<(), LexError> {
        match current {
            'a'..='z' | 'A'..='Z' | '_' | '0'..='9' => {
                self.state = State::Word {
                    start,
                    byte_len: byte_len + 1,
                }
            }
            _ => {
                self.finish_pending_state_and_reset()?;
                self.handle_char(current_index, current)?;
            }
        }
        Ok(())
    }

    fn handle_char_assuming_state_is_string(
        &mut self,
        current: char,
        start: ByteIndex,
        byte_len: usize,
    ) -> Result<(), LexError> {
        self.state = State::String {
            start,
            byte_len: byte_len + current.len_utf8(),
        };

        if current == '"' {
            self.finish_pending_state_and_reset()?;
        }

        Ok(())
    }

    fn handle_char_assuming_state_is_slash(
        &mut self,
        current_index: ByteIndex,
        current: char,
    ) -> Result<(), LexError> {
        if current == '/' {
            self.state = State::SingleLineComment;
            return Ok(());
        }

        self.finish_pending_state_and_reset()?;
        self.handle_char(current_index, current)
    }

    fn handle_char_assuming_state_is_dash(
        &mut self,
        current_index: ByteIndex,
        current: char,
    ) -> Result<(), LexError> {
        if current == '>' {
            let dash_index = ByteIndex(current_index.0 - '-'.len_utf8());
            self.out.push(Token::ThinArrow(dash_index));
            self.state = State::Main;
            return Ok(());
        }

        self.finish_pending_state_and_reset()?;
        self.handle_char(current_index, current)
    }

    fn handle_char_assuming_state_is_single_line_comment(
        &mut self,
        current: char,
    ) -> Result<(), LexError> {
        if current == '\n' {
            self.state = State::Main;
        }
        Ok(())
    }

    fn finish_pending_state_and_reset(&mut self) -> Result<(), LexError> {
        self.finish_pending_state()?;
        self.state = State::Main;
        Ok(())
    }

    fn finish_pending_state(&mut self) -> Result<(), LexError> {
        match self.state {
            State::Main => Ok(()),

            State::Word { start, byte_len } => {
                let word_src = &self.src[start.0..start.0 + byte_len];
                let Some(word) = parse_word(word_src, start) else {
                    return Err(LexError(start, ByteIndex(start.0 + byte_len)));
                };
                self.out.push(word);
                Ok(())
            }

            State::String { start, byte_len } => {
                let end = ByteIndex(start.0 + byte_len);
                let quote_exclusive_start = ByteIndex(start.0 + '"'.len_utf8());
                let quote_exclusive_end = ByteIndex(start.0 + byte_len - '"'.len_utf8());
                let quote_exclusive_string_src =
                    &self.src[quote_exclusive_start.0..quote_exclusive_end.0];
                let value = match get_string_value(quote_exclusive_string_src) {
                    Ok(v) => v,
                    Err(local_span) => {
                        return Err(LexError(
                            quote_exclusive_start + local_span.0,
                            quote_exclusive_start + local_span.1,
                        ));
                    }
                };
                self.out.push(Token::StringLiteral(StringLiteral {
                    value,
                    span: (start, end),
                }));
                Ok(())
            }
            State::Slash(start) => Err(LexError(start, ByteIndex(start.0 + '/'.len_utf8()))),
            State::Dash(start) => {
                self.out.push(Token::Dash(start));
                Ok(())
            }
            State::SingleLineComment => Ok(()),
        }
    }
}

fn parse_word(s: &str, start: ByteIndex) -> Option<Token> {
    if let Ok(val) = s.parse::<usize>() {
        return Some(Token::NumberLiteral(NumberLiteral {
            value: val,
            span: (start, ByteIndex(start.0 + s.len())),
        }));
    }

    match s {
        "_" => return Some(Token::Underscore(start)),

        "def" => return Some(Token::DefKw(start)),

        "match" => return Some(Token::MatchKw(start)),
        "fun" => return Some(Token::FunKw(start)),
        "For" => return Some(Token::ForKw(start)),

        "case" => return Some(Token::CaseKw(start)),
        "use" => return Some(Token::UseKw(start)),
        "end" => return Some(Token::EndKw(start)),
        "dec" => return Some(Token::DecKw(start)),

        _ => {}
    }

    if s.starts_with("Set") {
        let level = get_nonzero_number_after_prefix_but_return_zero_if_empty_string(s, "Set")?;
        return Some(Token::CapitalizedUniverseLiteral(
            CapitalizedUniverseLiteral {
                level,
                start,
                erasable: false,
            },
        ));
    }

    if s.starts_with("Prop") {
        let level = get_nonzero_number_after_prefix_but_return_zero_if_empty_string(s, "Prop")?;
        return Some(Token::CapitalizedUniverseLiteral(
            CapitalizedUniverseLiteral {
                level,
                start,
                erasable: true,
            },
        ));
    }

    if s.starts_with("set") {
        let level = get_nonzero_number_after_prefix_but_return_zero_if_empty_string(s, "set")?;
        return Some(Token::LowercaseUniverseLiteral(LowercaseUniverseLiteral {
            level,
            start,
            erasable: false,
        }));
    }

    if s.starts_with("prop") {
        let level = get_nonzero_number_after_prefix_but_return_zero_if_empty_string(s, "prop")?;
        return Some(Token::LowercaseUniverseLiteral(LowercaseUniverseLiteral {
            level,
            start,
            erasable: true,
        }));
    }

    if s.is_empty() {
        return None;
    }

    match s.chars().next().unwrap() {
        'a'..='z' | 'A'..='Z' | '_' => {
            return Some(Token::Ident(Ident {
                value: s.to_owned(),
                start,
            }))
        }
        _ => {}
    }

    None
}

/// - If `&s[prefix.len()..]` is empty, this function returns `Some(0)`.
///
/// - If `&s[prefix.len()..]` is a nonzero number `n` with no leading zeros,
///   this function returns `Some(n)`.
///
/// - Otherwise, this function returns `None`.
///
///   - As a corollary, if `&s[prefix.len()..]` is a number with extraneous leading zeros,
///     this function returns `None`.
fn get_nonzero_number_after_prefix_but_return_zero_if_empty_string(
    s: &str,
    prefix: &str,
) -> Option<usize> {
    let level_src = &s[prefix.len()..];
    if level_src.is_empty() {
        return Some(0);
    }

    if level_src.starts_with('0') {
        return None;
    }

    let Ok(level) = level_src.parse::<usize>() else {
        return None;
    };

    Some(level)
}
