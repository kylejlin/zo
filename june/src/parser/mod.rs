pub use generated_parser::parse;
mod generated_parser;

pub mod cst {
    pub use super::generated_parser::*;
}
