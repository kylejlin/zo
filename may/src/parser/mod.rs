pub use generated_parser::*;
mod generated_parser;

#[cfg(test)]
mod tests;
