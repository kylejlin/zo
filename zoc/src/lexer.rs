use crate::token::*;

pub struct LexError(ByteIndex, ByteIndex);

pub fn lex(s: &str) -> Result<Vec<Token>, LexError> {
    Lexer::new(s).lex()
}

struct Lexer<'a> {
    src: &'a str,
    out: Vec<Token>,
    state: State,
}

#[derive(Clone, Copy)]
enum State {
    Main,
    Word { start: ByteIndex, byte_len: usize },
    String { start: ByteIndex, byte_len: usize },
}

impl<'a> Lexer<'a> {
    fn new(src: &'a str) -> Self {
        Lexer {
            src,
            out: Vec::new(),
            state: State::Main,
        }
    }
}

impl Lexer<'_> {
    fn lex(mut self) -> Result<Vec<Token>, LexError> {
        for (current_index, curent) in self.src.char_indices() {
            self.handle_char(ByteIndex(current_index), curent)?;
        }
        self.finish_pending_state_and_reset()?;
        Ok(self.out)
    }

    fn handle_char(&mut self, current_index: ByteIndex, current: char) -> Result<(), LexError> {
        match self.state {
            State::Main => self.handle_char_assuming_state_is_main(current_index, current),
            State::Word { start, byte_len } => {
                self.handle_char_assuming_state_is_word(current_index, current, start, byte_len)
            }
            State::String { start, byte_len } => {
                self.handle_char_assuming_state_is_string(current, start, byte_len)
            }
        }
    }

    fn handle_char_assuming_state_is_main(
        &mut self,
        current_index: ByteIndex,
        current: char,
    ) -> Result<(), LexError> {
        match current {
            ' ' | '\t' | '\n' => {}
            '"' => {
                self.state = State::String {
                    start: current_index,
                    byte_len: '"'.len_utf8(),
                }
            }
            'a'..='z' | 'A'..='Z' | '_' | '0'..='9' => {
                self.state = State::Word {
                    start: current_index,
                    byte_len: 1,
                }
            }
            '(' => self.out.push(Token::LParen(current_index)),
            ')' => self.out.push(Token::RParen(current_index)),
            _ => {
                return Err(LexError(
                    current_index,
                    ByteIndex(current_index.0 + current.len_utf8()),
                ))
            }
        }
        Ok(())
    }

    fn handle_char_assuming_state_is_word(
        &mut self,
        current_index: ByteIndex,
        current: char,
        start: ByteIndex,
        byte_len: usize,
    ) -> Result<(), LexError> {
        match current {
            'a'..='z' | 'A'..='Z' | '_' | '0'..='9' => {
                self.state = State::Word {
                    start,
                    byte_len: byte_len + 1,
                }
            }
            _ => {
                self.finish_pending_state_and_reset()?;
                self.handle_char(current_index, current)?;
            }
        }
        Ok(())
    }

    fn handle_char_assuming_state_is_string(
        &mut self,
        current: char,
        start: ByteIndex,
        byte_len: usize,
    ) -> Result<(), LexError> {
        self.state = State::String {
            start,
            byte_len: byte_len + current.len_utf8(),
        };

        if current == '"' {
            self.finish_pending_state_and_reset()?;
        }

        Ok(())
    }

    fn finish_pending_state_and_reset(&mut self) -> Result<(), LexError> {
        self.finish_pending_state()?;
        self.state = State::Main;
        Ok(())
    }

    fn finish_pending_state(&mut self) -> Result<(), LexError> {
        match self.state {
            State::Main => Ok(()),

            State::Word { start, byte_len } => {
                let word_src = &self.src[start.0..start.0 + byte_len];
                let Some(word) = parse_word(word_src, start) else {
                    return Err(LexError(start, ByteIndex(start.0 + byte_len)));
                };
                self.out.push(word);
                Ok(())
            }

            State::String { start, byte_len } => {
                let end = ByteIndex(start.0 + byte_len);
                let quote_exclusive_start = ByteIndex(start.0 + '"'.len_utf8());
                let quote_exclusive_end = ByteIndex(start.0 + byte_len - '"'.len_utf8());
                let quote_exclusive_string_src =
                    &self.src[quote_exclusive_start.0..quote_exclusive_end.0];
                let value = match get_string_value(quote_exclusive_string_src) {
                    Ok(v) => v,
                    Err(local_span) => {
                        return Err(LexError(
                            quote_exclusive_start + local_span.0,
                            quote_exclusive_start + local_span.1,
                        ));
                    }
                };
                self.out.push(Token::String(StringLiteral {
                    value,
                    span: (start, end),
                }));
                Ok(())
            }
        }
    }
}

fn parse_word(s: &str, start: ByteIndex) -> Option<Token> {
    if let Ok(val) = s.parse::<usize>() {
        return Some(Token::Number(NumberLiteral {
            value: val,
            span: (start, ByteIndex(start.0 + s.len())),
        }));
    }

    match s {
        "ind" => return Some(Token::IndKw(start)),
        "vcon" => return Some(Token::VconKw(start)),
        "match" => return Some(Token::MatchKw(start)),
        "fun" => return Some(Token::FunKw(start)),
        "forall" => return Some(Token::ForallKw(start)),
        "nonrec" => return Some(Token::NonrecKw(start)),
        _ => {}
    }

    if s.starts_with("Type") {
        let level_src = &s["Type".len()..];
        if level_src.is_empty() {
            return None;
        }

        let has_extraneous_leading_zeros = level_src != "0" && level_src.starts_with('0');
        if has_extraneous_leading_zeros {
            return None;
        }

        let Ok(level) = level_src.parse::<usize>() else {
            return None;
        };

        return Some(Token::Universe(UniverseLiteral { level, start }));
    }

    None
}

use string_parser::get_string_value;
mod string_parser {

    use super::*;

    /// This function assumes the `quote_exclusive_string_src` argument
    /// does not have any double quotes.
    /// If `quote_exclusive_string_src` has double quotes,
    /// this function may produce an incorrect result.
    ///
    /// This function returns `Err(Some(span))` if it encounters
    /// an invalid escape sequence, where `span` is the location
    /// of the invalid escape sequence.
    ///
    /// This function returns `Err(None)` if it encounters
    /// an unterminated escape sequence.
    pub fn get_string_value(
        quote_exclusive_string_src: &str,
    ) -> Result<String, (ByteIndex, ByteIndex)> {
        StringParser::new(quote_exclusive_string_src).parse()
    }

    struct StringParser<'a> {
        quote_exclusive_string_src: &'a str,
        out: String,
        state: State,
    }

    #[derive(Clone, Copy)]
    enum State {
        Main,
        Escape { start: ByteIndex, byte_len: usize },
    }

    impl<'a> StringParser<'a> {
        fn new(quote_exclusive_string_src: &'a str) -> Self {
            Self {
                quote_exclusive_string_src,
                out: String::new(),
                state: State::Main,
            }
        }
    }

    impl StringParser<'_> {
        fn parse(mut self) -> Result<String, (ByteIndex, ByteIndex)> {
            for (current_index, current) in self.quote_exclusive_string_src.char_indices() {
                self.handle_char(ByteIndex(current_index), current)?;
            }

            match self.state {
                State::Escape { start, .. } => Err((
                    start,
                    ByteIndex(start.0 + self.quote_exclusive_string_src.len()),
                )),
                State::Main => Ok(self.out),
            }
        }

        fn handle_char(
            &mut self,
            current_index: ByteIndex,
            current: char,
        ) -> Result<(), (ByteIndex, ByteIndex)> {
            match self.state {
                State::Main => self.handle_char_assuming_state_is_main(current_index, current),
                State::Escape { start, byte_len } => self.handle_char_assuming_state_is_escape(
                    current_index,
                    current,
                    start,
                    byte_len,
                ),
            }
        }

        fn handle_char_assuming_state_is_main(
            &mut self,
            current_index: ByteIndex,
            current: char,
        ) -> Result<(), (ByteIndex, ByteIndex)> {
            match current {
                '{' => {
                    self.state = State::Escape {
                        start: current_index,
                        byte_len: '{'.len_utf8(),
                    };
                    Ok(())
                }
                _ => {
                    self.out.push(current);
                    Ok(())
                }
            }
        }

        fn handle_char_assuming_state_is_escape(
            &mut self,
            current_index: ByteIndex,
            current: char,
            start: ByteIndex,
            byte_len: usize,
        ) -> Result<(), (ByteIndex, ByteIndex)> {
            if current == '}' {
                let brace_exclusive_start = ByteIndex(start.0 + '{'.len_utf8());
                let brace_exclusive_end = current_index;
                self.push_escape_sequence(brace_exclusive_start, brace_exclusive_end)?;
                self.state = State::Main;
                return Ok(());
            }

            self.state = State::Escape {
                start: current_index,
                byte_len: byte_len + current.len_utf8(),
            };
            Ok(())
        }

        fn push_escape_sequence(
            &mut self,
            brace_exclusive_start: ByteIndex,
            brace_exclusive_end: ByteIndex,
        ) -> Result<(), (ByteIndex, ByteIndex)> {
            let invalid_escape_sequence_err = Err((brace_exclusive_start, brace_exclusive_end));

            let byte_len = brace_exclusive_end.0 - brace_exclusive_start.0;
            if byte_len < 3 {
                return invalid_escape_sequence_err;
            }

            let brace_exclusive_src =
                &self.quote_exclusive_string_src[brace_exclusive_start.0..brace_exclusive_end.0];
            if !brace_exclusive_src.starts_with("0x") {
                return invalid_escape_sequence_err;
            }

            let hex_start = ByteIndex(brace_exclusive_start.0 + "0x".len());
            let hex_src = &self.quote_exclusive_string_src[hex_start.0..brace_exclusive_end.0];
            let Ok(val) = u32::from_str_radix(hex_src, 16) else {
                return invalid_escape_sequence_err;
            };
            let Ok(val) = char::try_from(val) else {
                return invalid_escape_sequence_err;
            };
            self.out.push(val);
            Ok(())
        }
    }
}
