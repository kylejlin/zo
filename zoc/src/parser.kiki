start Expr

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
enum Expr {
    Ind(Ind)
    Vcon(Vcon)
    Match(Match)
    Fun(Fun)
    App(App)
    For(For)
    Deb($Number)
    Universe($Universe)
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct Ind {
    lparen: $LParen

    _: $IndKw
    type_: $Universe
    name: $String

    index_types_lparen: $LParen
    index_types: ZeroOrMoreExprs
    index_types_rparen: $RParen

    constructor_defs_lparen: $LParen
    constructor_defs: ZeroOrMoreVariantConstructorDefs
    constructor_defs_rparen: $RParen

    rparen: $RParen
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
enum ZeroOrMoreExprs {
    Nil
    Cons(ZeroOrMoreExprs Expr)
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
enum ZeroOrMoreVariantConstructorDefs {
    Nil
    Cons(ZeroOrMoreVariantConstructorDefs VariantConstructorDef)
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct VariantConstructorDef {
    lparen: $LParen

    param_types_lparen: $LParen
    param_types: ZeroOrMoreExprs
    param_types_rparen: $RParen

    index_args_lparen: $LParen
    index_args: ZeroOrMoreExprs
    index_args_rparen: $RParen

    rparen: $RParen
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct Vcon {
    lparen: $LParen

    _: $VconKw
    ind: Ind
    vcon_index: $Number

    rparen: $RParen
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct Match {
    lparen: $LParen

    _: $MatchKw
    matchee: Expr
    return_type: Expr

    cases_lparen: $LParen
    cases: ZeroOrMoreExprs
    cases_rparen: $RParen

    rparen: $RParen
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct Fun {
    lparen: $LParen

    _: $FunKw
    decreasing_index: NumberOrNonrecKw

    param_types_lparen: $LParen
    param_types: ZeroOrMoreExprs
    param_types_rparen: $RParen

    return_type: Expr
    return_val: Expr

    rparen: $RParen
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
enum NumberOrNonrecKw {
    Number($Number)
    NonrecKw($NonrecKw)
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct App {
    lparen: $LParen

    callee: Expr
    args: ZeroOrMoreExprs

    rparen: $RParen
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct For {
    lparen: $LParen

    _: $ForKw

    param_types_lparen: $LParen
    param_types: ZeroOrMoreExprs
    param_types_rparen: $RParen

    return_type: Expr

    rparen: $RParen
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
terminal Token {
    $LParen: crate::token::ByteIndex
    $RParen: crate::token::ByteIndex

    $IndKw: crate::token::ByteIndex
    $VconKw: crate::token::ByteIndex
    $MatchKw: crate::token::ByteIndex
    $FunKw: crate::token::ByteIndex
    $ForKw: crate::token::ByteIndex

    $NonrecKw: crate::token::ByteIndex

    $Number: crate::token::NumberLiteral
    $String: crate::token::StringLiteral
    $Universe: crate::token::UniverseLiteral
}
