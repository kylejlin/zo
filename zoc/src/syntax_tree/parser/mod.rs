mod generated_parser;
pub use generated_parser::*;

#[cfg(test)]
mod tests;
