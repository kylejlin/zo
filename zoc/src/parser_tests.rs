use crate::lexer::lex;
use crate::parser::*;

#[test]
fn deb_0() {
    let src = r#"0"#;
    let tokens = lex(src).unwrap();
    let cst = parse(tokens).unwrap();
    insta::assert_debug_snapshot!(&cst);
}

#[test]
fn deb_1() {
    let src = r#"1"#;
    let tokens = lex(src).unwrap();
    let cst = parse(tokens).unwrap();
    insta::assert_debug_snapshot!(&cst);
}

#[test]
fn deb_2() {
    let src = r#"2"#;
    let tokens = lex(src).unwrap();
    let cst = parse(tokens).unwrap();
    insta::assert_debug_snapshot!(&cst);
}

#[test]
fn app_no_args() {
    let src = r#"(0)"#;
    let tokens = lex(src).unwrap();
    let cst = parse(tokens).unwrap();
    insta::assert_debug_snapshot!(&cst);
}

#[test]
fn app_1_arg() {
    let src = r#"(0 1)"#;
    let tokens = lex(src).unwrap();
    let cst = parse(tokens).unwrap();
    insta::assert_debug_snapshot!(&cst);
}

#[test]
fn ind() {
    let src = r#"
(
    ind

    Type0

    "Nat"

    // Index types
    ()

    // Variant constructor defs
    (
        (() ())
        ((0) ())
    )
)"#;
    let tokens = lex(src).unwrap();
    let cst = parse(tokens).unwrap();
    insta::assert_debug_snapshot!(&cst);
}

#[test]
fn ind_nonliteral_universe() {
    let src = r#"
(
    ind

    1

    "Nat"

    // Index types
    ()

    // Variant constructor defs
    (
        (() ())
        ((0) ())
    )
)"#;
    let tokens = lex(src).unwrap();
    let err = parse(tokens).unwrap_err();
    insta::assert_debug_snapshot!(&err);
}

#[test]
fn ind_nonliteral_name() {
    let src = r#"
(
    ind

    Type0

    2

    // Index types
    ()

    // Variant constructor defs
    (
        (() ())
        ((0) ())
    )
)"#;
    let tokens = lex(src).unwrap();
    let err = parse(tokens).unwrap_err();
    insta::assert_debug_snapshot!(&err);
}

#[test]
fn vcon() {
    let src = r#"
(
    vcon

    (
        ind

        Type0

        "Nat"

        // Index types
        ()

        (
            (() ())
            ((0) ())
        )
    )

    // Variant constructor index - THIS MUST BE A NUMBER LITERAL.
    0
)"#;
    let tokens = lex(src).unwrap();
    let cst = parse(tokens).unwrap();
    insta::assert_debug_snapshot!(&cst);
}

#[test]
fn vcon_nonliteral_ind() {
    let src = r#"
(
    vcon

    1

    // Variant constructor index - THIS MUST BE A NUMBER LITERAL.
    0
)"#;
    let tokens = lex(src).unwrap();
    let err = parse(tokens).unwrap_err();
    insta::assert_debug_snapshot!(&err);
}

#[test]
fn vcon_nonliteral_vcon_index() {
    let src = r#"
(
    vcon

    (
        ind

        Type0

        "Nat"

        // Index types
        ()

        (
            (() ())
            ((0) ())
        )
    )

    (2 3)
)"#;
    let tokens = lex(src).unwrap();
    let err = parse(tokens).unwrap_err();
    insta::assert_debug_snapshot!(&err);
}
