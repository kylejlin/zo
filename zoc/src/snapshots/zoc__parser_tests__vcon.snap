---
source: src/parser_tests.rs
expression: "&cst"
---
Vcon(
    Vcon {
        lparen: ByteIndex(
            1,
        ),
        ind: Ind {
            lparen: ByteIndex(
                17,
            ),
            type_: UniverseLiteral {
                level: 0,
                start: ByteIndex(
                    40,
                ),
            },
            name: StringLiteral {
                value: "Nat",
                span: (
                    ByteIndex(
                        55,
                    ),
                    ByteIndex(
                        60,
                    ),
                ),
            },
            index_types_lparen: ByteIndex(
                93,
            ),
            index_types: Nil,
            index_types_rparen: ByteIndex(
                94,
            ),
            constructor_defs_lparen: ByteIndex(
                105,
            ),
            constructor_defs: Cons(
                Cons(
                    Nil,
                    VariantConstructorDef {
                        lparen: ByteIndex(
                            119,
                        ),
                        param_types_lparen: ByteIndex(
                            120,
                        ),
                        param_types: Nil,
                        param_types_rparen: ByteIndex(
                            121,
                        ),
                        index_args_lparen: ByteIndex(
                            123,
                        ),
                        index_args: Nil,
                        index_args_rparen: ByteIndex(
                            124,
                        ),
                        rparen: ByteIndex(
                            125,
                        ),
                    },
                ),
                VariantConstructorDef {
                    lparen: ByteIndex(
                        139,
                    ),
                    param_types_lparen: ByteIndex(
                        140,
                    ),
                    param_types: Cons(
                        Nil,
                        Deb(
                            NumberLiteral {
                                value: 0,
                                span: (
                                    ByteIndex(
                                        141,
                                    ),
                                    ByteIndex(
                                        142,
                                    ),
                                ),
                            },
                        ),
                    ),
                    param_types_rparen: ByteIndex(
                        142,
                    ),
                    index_args_lparen: ByteIndex(
                        144,
                    ),
                    index_args: Nil,
                    index_args_rparen: ByteIndex(
                        145,
                    ),
                    rparen: ByteIndex(
                        146,
                    ),
                },
            ),
            constructor_defs_rparen: ByteIndex(
                156,
            ),
            rparen: ByteIndex(
                162,
            ),
        },
        vcon_index: NumberLiteral {
            value: 0,
            span: (
                ByteIndex(
                    235,
                ),
                ByteIndex(
                    236,
                ),
            ),
        },
        rparen: ByteIndex(
            237,
        ),
    },
)
